Final config

zh23jemu opened the pull request.

zh23jemu committed:
Update .gitignore, aspell_exceptions & pylint

zh23jemu committed:
Create configs.config module with Conf class

Add jsonschema to requirements file.
Define APPLICATION_DIR in global_vars module.

zh23jemu committed:
Add ApplicationConf class to configs.config module

zh23jemu committed:
Add ProjectConf class to configs.config module

zh23jemu committed:
Add OperatorConf class to configs.config module

LegenJCdary commented:
After this is implemented, please close #5, #10, #13 and #22.

zh23jemu committed:
Add MergedConf class to configs.config

Define CONFIG_RULES in global_vars, create configs.templates.final.

LegenJCdary commented:
2 general advices:
- Try to create small commits, even tiny ones (at least until you push for review) - treat them as building blocks for your PR.
- Use `git rebase -i HEAD~NN` frequently and get comfortable with that tool. With tiny blocks/commits you can rearrange your building/PR easily with e.g. fixup (correct and commit each problem separately at the end of the branch and then fix to different points in your branch) or squash (commit each file/class/function/code block separately and then squash if they are functionally/logically related).
- Run CI jobs locally to solve problems as soon as possible. With small commits, comfort in using rebase command and utilizing developer scripts you can easily run each CI test on your dev environment. A good practice would be e.g.
```
(before commit or during rebase)
1. make some changes
2. ./tools/prep_test.sh
3. pylint + pytest -> make corrections -> pylint + pytest
4. ./tools/prep_commit.sh
5a. git commit
5b. git add && git rebase --continue
```

LegenJCdary commented on src/modules/configs/schemas/application.py:
```diff
@@ -0,0 +1,89 @@
+application = {
```
Schemas related to each conf type should be committed (fixed up to or squashed) with commits introducing classes for these conf types e.g. `application` with `ApplicationConf`.

LegenJCdary replied:
Also, please rename to something generic like `schema` or `template`.

LegenJCdary commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 pylint
```
Fixup to commit where jsonschema is first used.

LegenJCdary commented on src/modules/global_vars/__init__.py:
```diff
@@ -74,5 +51,28 @@
         "mode": "priority",
         "priority": ("operator", "project", "application")
+    },
+    "naming_log_file": {
```
Why change alphabetical order?

LegenJCdary replied:
On the other hand... order here is up to you.

LegenJCdary commented on src/modules/global_vars/__init__.py:
```diff
@@ -74,5 +51,28 @@
+        "inclusive": ("operator", "project", "application"),
+        "default": {}
+    },
+    "exclude_devices": {
```
Field names should be adjusted to final template here: b5190492921025f4806e42db88fc8f0b8324aa6e.

LegenJCdary commented on src/modules/global_vars/__init__.py:
```diff
@@ -3,11 +3,11 @@
 CONFIG_RULES = {
```
`CONFIG_RULES` corrections should be fixed up to b5190492921025f4806e42db88fc8f0b8324aa6e.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -59,88 +54,83 @@ class ApplicationConf(Conf):
+        self.pre_final = self.create_final_conf()
+        self.final = self.validate_final_conf()
+
+    def validate_final_conf(self):
```
This should be either inherited from the Conf class or the validation function should be extracted outside of Conf class (those will land in the `__init__` most likely but we will get to that later).

zh23jemu committed:
Create test functions for Conf and its sub-classes

LegenJCdary commented on tests/data/configs/application.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "containers": {
+    "sata": {
+      "device_path": "/dev/sda",
```
Should be `/dev`, sda is already a device and we want root path.

LegenJCdary commented on src/modules/configs/schemas/application.py:
```diff
@@ -0,0 +1,89 @@
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "properties": {
+        "containers": {
```
I think destination and permission objects should be added here too.

LegenJCdary replied:
You can think about (or plan for future via issues/milestones) defining smaller objects and using jsonschema import to re-use existing definitions e.g. `notify` object will be used in all 3 schemas.

LegenJCdary commented on tests/data/configs/application.json:
```diff
@@ -0,0 +1,24 @@
+    "kuba@ingestation.com"
+  ],
+  "exclude": [
+    "DISKSN02"
```
e.g. WD-WCC4E5XB5R5F

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+from tests.data.configs import validate_conf, err_msg, config_rule
+
+current_path = os.path.join(os.path.dirname(__file__))
+test_data_path = os.path.join(os.path.dirname(os.path.dirname(current_path)),
```
As class variable if required.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+    def test_get_conf_path(self, conf_path, conf_file, real_conf_path):
+        try:
+            if conf_file == "application.json":
+                conf = ApplicationConf("")
```
try to parametrize that too - might just work

zh23jemu replied:
tried to split to seperate test functions

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+    """Class to test functions of Conf class"""
+
+    @pytest.mark.parametrize("conf_path, conf_file, real_conf_path", [
+        [os.path.join(test_data_path, "application.json"), "application.json",
```
You can move this list to class variable e.g. `conf_list` and use it in fixture as `TestConf.conf_list`. If data gets too bulky you can create locally `TestConfData` and add only variables there.

zh23jemu replied:
tried to split to separate test functions

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+    """Class to test functions of Conf class"""
+
+    @pytest.mark.parametrize("conf_path, conf_file, real_conf_path", [
+        [os.path.join(test_data_path, "application.json"), "application.json",
```
`os.path.join` is doubled

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@pytest.mark.parametrize("conf, schema_file, err_msg", [
+    [validate_conf.application.get("unrecognized_key"), application,
```
Each of those "getattrs" could be done in the function, would clear out the code.

LegenJCdary commented on tests/data/configs/validate_conf.py:
```diff
@@ -0,0 +1,192 @@
+            }
+        }
+    },
+    "successful_validation": {
```
Imo successful case should be separated from failed one, both in test data and in code - few more lines but much clearer picture.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,86 @@
+    [validate_conf.project.get("successful_validation"), project, ""],
+    [validate_conf.operator.get("successful_validation"), operator, ""],
+])
+def test_validate_dict(conf, schema_file, err_msg):
```
As stated above, this should be divided into e.g. `test_validate_json_passed()` and `test_validate_json_failed()`. When we run pytest (either manually or CI) we want to see e.g. 20 dots for test_validate_json_passed.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -81,3 +81,120 @@ class OperatorConf(Conf):
+                    "Order is undefined. Rule mode order needs to be defined for rule mode.")
+            if not value.get(mode):
+                raise Exception("Mode is undefined")
+            final_conf[key] = self.get_final_value(key)
```
`final_conf` values need to be mapped back onto `final.template` and then this `mapped_final_conf` needs to be validated with proper schema. Not sure if merged conf (final_conf) needs to be validated.

LegenJCdary replied:
Quick solution (feel free to implement your own): split final_conf key by "_" and iterate .get() on final.template until you get not-None return.

LegenJCdary replied:
If you are struggling with that check out 6f36b684504aa5e70cee970c8251af977b3fc4e4. Not gonna lie, it took me few hours, mainly because I'm stubborn and bad with debugger, so I did most debugging with `print`s in Linux terminal (bad idea). Anyway, code is rather ugly and not optimized but it does the job. If you decide to use it, please rename variables, functions, remove `print` statements etc.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -81,3 +81,120 @@ class OperatorConf(Conf):
+                value_count += 1
+                effective_value = value
+            if value_count > 1:
+                raise Exception(f"{merged_key} is defined in more than one config.")
```
return config names too

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -81,3 +81,120 @@ class OperatorConf(Conf):
+            value = self.merged_conf[conf].get(merged_key)
+            if value is not None:
+                return value
+        return None
```
We should probably log warning here for now (not sure how we will handle priority mode with env variables) and then return validation error when variable is undefined in final.

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,62 @@
+    return ''.join([random.choice(letters) for _ in range(length)])
+
+
+def validate_dict(conf: dict, schema_file) -> None:
```
I kept it here and tried to raise another ValidationError in parse_conf(), which can include conf_type in error message.

zh23jemu replied:
And re-raise the ValidationError from the previous one, which is recommended by pylint XD

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,62 @@
+
+def validate_dict(conf: dict, schema_file) -> None:
+    if validate(instance=conf, schema=getattr(schema_file, "schema")) is not None:
+        raise Exception("Config file validation error.")
```
Invalid section of json is included in exception message.

zh23jemu commented on src/modules/configs/schemas/application.py:
```diff
@@ -0,0 +1,89 @@
+                }
+            },
+            "additionalProperties": False,
+            "required": [
```
what does imo mean? XD

zh23jemu replied:
i see. operator has the highest priority for log_file and codename.

zh23jemu commented on src/modules/configs/schemas/project.py:
```diff
@@ -0,0 +1,129 @@
+                        },
+                        "mode": {
+                            "type": "string",
+                            "maxLength": 4096
```
used regex to fix the length

zh23jemu committed:
Create test functions for validate_json()

Update github pytest workflow.
Add configuration files for pytest.
Add functions in tests/utils/__init__.py.
Add configuration schema files.

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -81,3 +81,120 @@ class OperatorConf(Conf):
+    def validate_final_conf(self):
+        if validate_dict(self.pre_final, final_schema) is None:
+            return self.pre_final
+        raise Exception("Final config validation error.")
```
moved it to create_final_conf() and added raise exception from ValidationError,

zh23jemu commented on src/modules/global_vars/__init__.py:
```diff
@@ -0,0 +1,78 @@
+CONFIG_RULES = {
+    "containers_sata_device_path": {
+        "mode": "exclusive",
+        "exclusive": ("application",)
```
nope, i tested before. skipping commas will make it a string not a tuple.

zh23jemu committed:
Add test functions for MergedConf class

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+
+        try:
+            validate_json(conf, self.schema_file)
+        except ValidationError as ex:
```
Imo `self.logger.error(exc)` and then logging critical (using `ColourFormatter`) looks better. 
![image](https://github.com/LegenJCdary/ingeSTation/assets/75223465/5fa9fcf3-f136-41fb-918a-68358d82db43)

zh23jemu replied:
Will try to use the Loggers class

LegenJCdary replied:
Just pass from `main` to `MergedConf` and then to other conf classes.

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+    def parse_conf(self) -> Union[bool, dict]:
+        conf = self.load_conf()
+
+        try:
```
I think this exc handling should be done directly in `validate_json()` - this way you don't have to repeat it in `MergedConf.create_final_conf()`.

zh23jemu replied:
ok, tested that can get config type from schema_file name

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+            if os.path.isfile(conf_path):
+                return conf_path
+            raise FileNotFoundError(f"[CRITICAL]: {self.conf_type.capitalize()}"
+                                    " configuration could not be obtained.")
```
*... could not be obtained/found at {os.path.dirname(self.conf_path)}*

LegenJCdary replied:
Remember to add whitespace between a path and fullstop - nice gesture towards terminal users (double-clicking in Putty marks surrounding interpunction).

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+
+        try:
+            validate_json(conf, self.schema_file)
+        except ValidationError as ex:
```
Raising exception is the preferred Python solution and I am not trying to undermine that. However, this application is targeting ingest Operators and (from my experience) they tend to be more on the user than developer side of things, so we should keep UX in mind. Best solution would be to create a child class in `modules.outputs` that would colour our exceptions and perhaps limit tracebacks in some cases (most users only read the last line anyway) - similar to `modules.outputs.loggers.ColourFormatter`. For sure this is a task for another MR (#44).

zh23jemu replied:
you are right. let me think about it

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+            validate_json(conf, self.schema_file)
+        except ValidationError as ex:
+            raise ValidationError(
+                f"[CRITICAL]: {self.conf_type.capitalize()} config validation failed.") \
```
We should mention sth like *Review application config at {self.conf_path} and correct errors.*

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,91 @@
+
+from jsonschema import ValidationError
+
+sys.path.append(os.getcwd())
```
We shouldn't modify sys.path in our final product.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,91 @@
+
+from ingestation.modules.configs.config import validate_json
+from ingestation.modules.configs.schemas import application, project, operator
+from tests.data.configs import validate_json_passed, validate_json_failed, err_msg
```
Such import simply won't work after install. We need to define a new package, update setup.cfg (example below) and adjust developer tools.
```
[options]
package_dir =
    ingestation=src
    ingestation_test=tests
packages =
    ingestation
    ingestation.modules
    ingestation.modules.configs
    ingestation.modules.configs.schemas
    ingestation.modules.configs.templates
    ingestation.modules.global_vars
    ingestation.modules.misc
    ingestation.modules.outputs
    ingestation_test
    ingestation_test.data
    ingestation_test.data.configs
```

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -1,10 +1,16 @@
 
-from ingestation.modules.global_vars import APPLICATION_DIR
+
+def generate_string(length):
```
Such functions should go to `tests/utils/__init__.py`.

LegenJCdary commented on tests/data/configs/application.json:
```diff
@@ -0,0 +1,24 @@
+  "resources": {
+    "thread_count": 1
+  },
+  "workdir": "/var/log/",
```
Remove os.sep

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -0,0 +1,91 @@
+from ingestation.modules.configs.schemas import application, project, operator
+from tests.data.configs import validate_json_passed, validate_json_failed, err_msg
+
+
```
Let's aggregate all those functions under a class e.g. `TestAllConfTypes`. Then, you can move parametrization to class-level e.g.
```
@pytest.mark.parametrize("conf_type,schema", [
    ["application", application],
    ["project", project],
    ["operator", operator]
])
class TestAllConfigTypes:

    def test_validate_json_unrecognized_key(self, conf_type: str, schema):
        full_conf = getattr(validate_json_failed, conf_type)
        conf = full_conf.get("unrecognized_key")

        with pytest.raises(ValidationError) as ex:
            validate_json(conf, schema)
        assert err_msg.validate_json.get("unrecognized_key") in str(ex.value)
```

LegenJCdary replied:
Besides `tests.modules.configs.test_config.TestConf` should come first in this commit. `TestAllConfTypes` should be committed only after each of their code is implemented.

LegenJCdary replied:
This work too:
```
@pytest.mark.parametrize("conf_type,schema", [
    ["application", application],
    ["project", project],
    ["operator", operator]
])
class TestAllConfigTypes:

    @pytest.mark.parametrize("validation_error", ["invalid_data_type", "mandatory_key",
        "too_long_value", "unrecognized_key"])
    def test_validate_json_failed(self, conf_type: str, schema, validation_error: str):
        full_conf = getattr(validate_json_failed, conf_type)
        conf = full_conf.get(validation_error)

        with pytest.raises(ValidationError) as ex:
            validate_json(conf, schema)
        assert err_msg.validate_json.get(validation_error) in str(ex.value)

    def test_validate_json_non_mandatory_key(self, conf_type: str, schema):
        full_conf = getattr(validate_json_passed, conf_type)
        conf = full_conf.get("non_mandatory_key")

        assert validate_json(conf, schema) is None

    def test_validate_json_passed(self, conf_type: str, schema):
        full_conf = getattr(validate_json_passed, conf_type)
        conf = full_conf.get("validation_passed")

        assert validate_json(conf, schema) is None
```

LegenJCdary commented on src/modules/configs/templates/application.py:
```diff
@@ -0,0 +1,19 @@
+config = {
```
This commit can be dropped, you are removing those files in the same MR.

LegenJCdary commented on tests/data/configs/err_msg.py:
Please rename to `error_messages` - it can be abbreviated upon import.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -89,3 +91,33 @@ def test_validate_json_validation_passed(conf_type: str, schema):
+    test_data_path = os.path.join(os.path.dirname(os.path.dirname(current_path)),
+                                  "data", "configs")
+
+    @pytest.mark.parametrize("conf_type", ["application", "project", "operator"])
```
```
    @pytest.mark.parametrize("conf_type,conf", [("application", ApplicationConf("")),
                                                ("project", ProjectConf("")), ("operator",
                                                                               OperatorConf(""))])
    def test_get_conf_path_application_dir(self, conf_type, conf):
        assert conf.get_conf_path("") == os.path.join(APPLICATION_DIR, f"{conf_type}.json")
```

LegenJCdary commented on src/modules/configs/schemas/final.py:
This schema is for `merged_conf`, not `final_conf`. Final conf schema should be a combination of all singular (app, project, operator) schemas. We can leave this schema and validate `merged_conf` additionally (prior to creating and validating `final_conf`) but currently I don't see any benefit in that.

zh23jemu replied:
yes, i removed the old merged_conf version

LegenJCdary commented on src/modules/configs/templates/final.py:
```diff
@@ -0,0 +1,57 @@
+# All fields are required!
```
All comments should be removed, this info can be obtained from schema file.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -121,3 +120,134 @@ def test_get_conf_path_file_not_found(self):
+        [merge_conf.dct2, merge_conf.expect2],
+        [merge_conf.dct3, merge_conf.expect3],
+    ])
+    def test_merge_conf(self, dct, expect):
```
This test is alright, although you could define more meaningful variable names (top-level only i.e. `dct1`, `expect1` etc.).

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -121,3 +120,134 @@ def test_get_conf_path_file_not_found(self):
+        [merge_conf.dct3, merge_conf.expect3],
+    ])
+    def test_merge_conf(self, dct, expect):
+        conf = MergedConf(Loggers({"debug": True}))
```
Why do we need debug here?

zh23jemu replied:
looks Loggers() only need a dict here, let me test again

LegenJCdary replied:
Ahhh, you are right! `Loggers.add_console_handler()` will throw KeyError if `debug` is undefined. I will fix that right away, so that empty dict is acceptable.

LegenJCdary replied:
debugs should be removed now, the issue is solved.

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -121,3 +120,134 @@ def test_get_conf_path_file_not_found(self):
+    @pytest.mark.parametrize("merged_key, rule_type", [
+        ["resources_thread_count", "exclusive_over_defined"]
+    ])
+    def test_apply_config_rule_exclusive_over_defined(self, merged_key, rule_type):
```
We should really check for conflicts between configs not internally. If we are expecting duplicates within each config, first of all we should handle that in the source code (perhaps using jsonschema or https://github.com/dmeranda/demjson).

LegenJCdary commented on tests/modules/configs/test_config.py:
```diff
@@ -121,3 +120,134 @@ def test_get_conf_path_file_not_found(self):
+    @pytest.mark.parametrize("merged_key, rule_type", [
+        ["resources_thread_count", "exclusive_undefined"]
+    ])
+    def test_apply_config_rule_exclusive_undefined(self, merged_key, rule_type):
```
`exclusive_undefined` `rule_type` is currently undefined. Same as for previous tests, you should try to create complete configs with single element altered/missing/invalid.

LegenJCdary replied:
Same as for previous tests, you should try to create complete configs with single element altered/missing/invalid... but that's for another day. Make those available tests work and create an issue for config tests improvement.

zh23jemu commented on tests/data/configs/application.json:
for now I think alphabetical order is ok

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,48 @@
+        return conf
+
+    def load_conf(self) -> dict:
+        with open(self.conf_path, "r", encoding="utf-8") as fh:
```
will add that

zh23jemu replied:
added in issue #44

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -81,3 +81,120 @@ class OperatorConf(Conf):
+        self.operator = OperatorConf("").parse_conf()
+
+        self.merged_conf = {
+            "cli": {},
```
So we need a cli_options parameter for __init__() of MergedConf here?

LegenJCdary replied:
Yes

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -75,3 +86,140 @@ class OperatorConf(Conf):
+        2. Return value if defined in any
+        3. If undefined -> return None
+        """
+        order = CONFIG_RULES.get(merged_key)["exclusive"]
```
you mean this place?
`value, effective_value = None, None`

LegenJCdary replied:
In this function you add blank line in-between local variable declarations and `return` without prior blank line. In the following function there is no blank line after var declarations but there is a blank line before `return`. In the 3rd function there is no blank line after vars and no blank line before `return`. Just try to stay consistent - eyes/brain adapt to some patterns and it's easier to read if patterns are consistent.

zh23jemu replied:
that's very good norm. will improve in following codes.

zh23jemu commented on src/modules/configs/schemas/application.py:
```diff
@@ -0,0 +1,158 @@
+                        }
+                    },
+                    "additionalProperties": False,
+                    "required": [
```
looks it means they are sub-properties of primary property.  we cannot define a application config like this:
```
   "permissions": {
        "primary": {
        }
    },
```

LegenJCdary replied:
Maybe let's require mode only? Makes more sense to require it than uid/gid - we should allow to inherit from user runtime at some point (initial use case is sudo mode anyway).

zh23jemu commented on src/modules/configs/schemas/application.py:
```diff
@@ -0,0 +1,158 @@
+        }
+    },
+    "additionalProperties": False,
+    "required": ["containers"]
```
we need this because it's the exclusive rule
```
    "containers_sata_device_path": {
        "mode": "exclusive",
        "exclusive": ("application",)
    },
    "containers_sata_mount_path": {
        "mode": "exclusive",
        "exclusive": ("application",)
    },
```

LegenJCdary replied:
It's indirectly required... We could postpone validation of that key until final conf but it's inefficient. My bad, I believe I was thinking about adding those to some other conf.

zh23jemu commented on src/modules/configs/schemas/project.py:
```diff
@@ -0,0 +1,130 @@
+                        },
+                        "mode": {
+                            "type": "string",
+                            "pattern": "^0o[1-7][0-7]{2}$"
```
`"^0o[0-7]{3}$"` ?

LegenJCdary replied:
Both of us are somewhat wrong.

- If we want to force certain "minimal" mode I would suggest `[4-7]` on the owner - imo read is more important than x...
- ...but I don't think we need to force that. If an operator wants `0000` on his files - be my guest. 😆 Root user will still be able to access them.
- In the end I think we need `{4}` to allow special mode setting (suid, sgid etc.) - I forgot about that initially.

zh23jemu committed:
Add unmerged_conf()

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -176,6 +204,30 @@ def merge_conf(self, dct):
+                    new_ex_key = str.replace(ex_key, "_", "-")
+                    key = str.replace(key, ex_key, new_ex_key)
+                    break
+            sep_key = str.split(key, "_")
```
`separated_key` or `split_key`

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -102,10 +102,40 @@ def __init__(self, logger: Loggers):
+        return lst
+
+    @staticmethod
+    def add_to_dict(dct, keys, value):
```
`add_values_in_dict` or `add_values_to_dict_keys`

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -102,10 +102,40 @@ def __init__(self, logger: Loggers):
-    def create_final_conf(self):
-        final_conf = self.merge_conf(final.template)
+    @staticmethod
+    def extract_keys(dct):
```
`extract_dict_keys`, `extract_keys_from_dict`, `extract_final_schema_keys` or `extract_keys_from_final_schema`

LegenJCdary reviewed (changes requested):
Looks good, short tests were successful. Sorry about nitpicking var/func names but initially the code was a bit incomprehensible to me.

zh23jemu commented on tests/data/configs/validate_json_failed.py:
maybe just merge validate_json_passed.py and validate_json_failed.py to one file allconf.py, because I need to use conf_type to get different test data.

zh23jemu committed:
Add packages to setup.cfg

LegenJCdary commented on tests/utils/__init__.py:
Please mention in the 9e13ee414f919b33b677a1c2e1a8023f67f523f9 commit body.

LegenJCdary commented on tests/configs/pytest/application.json:
Please mention in the 9e13ee414f919b33b677a1c2e1a8023f67f523f9 commit message.

LegenJCdary commented on tests/data/configs/validate_json_failed.py:
```diff
@@ -0,0 +1,103 @@
+            "primary": {
+                "uid": 0,
+                "gid": 0,
+                "mode": "0o700"
```
`0o0700`

LegenJCdary commented on tests/data/configs/validate_json_failed.py:
```diff
@@ -0,0 +1,103 @@
+# pylint: disable=duplicate-code
+
+from ingestation_test.utils import generate_string
+
```
2 blank lines

LegenJCdary commented on src/modules/configs/schemas/project.py:
```diff
@@ -0,0 +1,130 @@
+schema = {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "properties": {
```
_According to CONFIG_RULES key containers should also be here._
This still needs to be completed.

zh23jemu replied:
do we need containers here in project.py? i found there's only application in CONFIG_RULES here
```
    "containers_sata_device_path": {
        "mode": "exclusive",
        "exclusive": ("application",)
    },
    "containers_sata_mount_path": {
        "mode": "exclusive",
        "exclusive": ("application",)
    },
```

LegenJCdary commented on tests/configs/aspell_exceptions:
```diff
@@ -1,8 +1,12 @@
 abstractproperty
-analyse
+acl
```
`acl` still here...

LegenJCdary commented on src/main.py:
```diff
@@ -2,6 +2,7 @@
 from ingestation.modules.misc.utils import InitParams
 from ingestation.modules.outputs.loggers import Loggers
 from ingestation.modules.outputs import blocks
+from ingestation.modules.configs.config import MergedConf
```
should be sorted alphabetically

LegenJCdary commented on src/modules/configs/config.py:
```diff
@@ -25,6 +28,14 @@ def validate_json(conf: dict, schema_file) -> None:
             from ex
 
 
+class ConfigRulesError(Exception):
```
Shouldn't it actually `raise Exception` instead of `pass`?

zh23jemu replied:
nope, i refer to this article.  and it works fine
[https://realpython.com/python-raise-exception/#choosing-the-exception-to-raise-built-in-vs-custom](url)

LegenJCdary reviewed (changes requested):
Rough rebase order (imo) below.

05daa50 Create configs.config module with Conf class
d912ac2 Create test functions for Conf and its sub-classes
feb5fae Add ApplicationConf class to configs.config module
6f1d338 Add ProjectConf class to configs.config module
7233f90 Add OperatorConf class to configs.config module
9e13ee4 Create test functions for validate_json()
DROP -> 245c93c Templates for Application, Project & Operator
	29caf4c Add MergedConf class to configs.config
	SQUASH to 29caf4c -> 84701e6 Add unmerged_conf()
0defd2a Add test functions for MergedConf class
73f1015 Add packages to setup.cfg
8e8c958 Update .gitignore, aspell_exceptions & pylint

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -2,9 +2,8 @@
-from jsonschema import validate, ValidationError
-
 from ingestation.modules.global_vars import APPLICATION_DIR
+from jsonschema import validate, ValidationError
```
cannot figure it out. which change do you mean?
```
- from jsonschema import validate, ValidationError

  from ingestation.modules.global_vars import APPLICATION_DIR
+from jsonschema import validate, ValidationError
```

LegenJCdary replied:
`jsonschema` import should be above `ingestation` import in final version. Common import order: Python builtins -> external project packages (e.g. jsonschema, pylint - pretty much all we have in `requirements.txt`) -> current project packages.

zh23jemu commented on src/modules/configs/config.py:
```diff
@@ -0,0 +1,56 @@
+from ingestation.modules.global_vars import APPLICATION_DIR
+
+
+def convert_schema_name(schema_file):
```
the schema_file here is a file type, need to convert to string XD

zh23jemu commented on tests/data/configs/application.json:
nope, used in test_get_conf_path_file_found() to test when config file path is specified in get_conf_path()

LegenJCdary commented:
Follow up in #41 and #51.
